### PR TITLE
FEATURE: Reactions!

### PR DESCRIPTION
**Pull request by markvanlan**
Screen Shot 2021-12-02 at 2 39 13 PM

Screen Shot 2021-12-02 at 2 39 22 PM

Screen Shot 2021-12-02 at 2 39 30 PM

Screen Shot 2021-12-02 at 2 39 41 PM

Screen Shot 2021-12-02 at 2 39 48 PM

Screen Shot 2021-12-02 at 2 39 54 PM

### CODE REVIEW

File: app/serializers/chat_base_message_serializer.rb [hunk 1]
```diff
@@ -11,13 +11,35 @@ class ChatBaseMessageSerializer < ApplicationSerializer
...
+        count: reactions.count,
+        users: ActiveModel::ArraySerializer.new(users, each_serializer: BasicUserSerializer).as_json,
+        reacted: users_reactions.include?(emoji)
+      }
+    end
+    reactions_hash
+  end
+
+  def include_reactions?
+    object.reactions.any?
```
Comment (CvX) [inline app/serializers/chat_base_message_serializer.rb:36]: Could you double check this doesn't cause an extra `SELECT 1 FROM…` query?

Comment (markvanlan) [reply app/serializers/chat_base_message_serializer.rb:36]: I did, yes!

File: assets/javascripts/discourse/components/chat-message-reaction.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import Component from "@ember/component";
+import discourseComputed from "discourse-common/utils/decorators";
+import { bind } from "discourse-common/utils/decorators";
```
Comment (CvX) [inline assets/javascripts/discourse/components/chat-message-reaction.js:3]:
```suggestion
import discourseComputed, { bind } from "discourse-common/utils/decorators";
```

😛

**Review by CvX — APPROVED** [review]

File: assets/javascripts/discourse/templates/components/tc-message.hbs [hunk 3]
```diff
@@ -21,6 +27,12 @@
         {{#unless message.staged}}
           <div class="tc-msgactions-hover">
             <div class="tc-msgactions">
+              {{d-button
+                class="btn-flat react-btn"
+                action=(action "startReaction" ".react-btn" SHOW_LEFT)
```
Comment (pmusaraj) [inline assets/javascripts/discourse/templates/components/tc-message.hbs:32]: Feels a little unusual to pass `.react-btn`, both here, and in the `.chat-message-react-btn` below.

File: assets/javascripts/discourse/components/tc-message.js [hunk 4]
```diff
@@ -158,6 +199,170 @@ export default Component.extend({
...
+  hasReactions(reactions) {
+    return Object.values(reactions).some((r) => r.count > 0);
+  },
+
+  @action
+  showUsersList(reaction) {
+    let list;
+    let usernames = reaction.users.map((u) => u.username).join(", ");
+    if (reaction.reacted) {
+      if (reaction.count === 1) {
```
Comment (pmusaraj) [inline assets/javascripts/discourse/components/tc-message.js:212]: I would find a `switch case` statement more readable here.

File: assets/javascripts/discourse/components/tc-message.js [hunk 5]
```diff
@@ -158,6 +199,170 @@ export default Component.extend({
...
+      "chat-message:reaction-picker-opened",
+      this.message.id
+    );
+
+    schedule("afterRender", () => {
+      this._repositionEmojiPicker(btnSelector, position);
+    });
+  },
+
+  _repositionEmojiPicker(btnSelector, position) {
```
Comment (pmusaraj) [inline assets/javascripts/discourse/components/tc-message.js:261]: This repositioning of the emoji picker feels a little fragile. Could we use Popper instead? 

Also, not sure how this currently works on mobile, you probably went something full width there.

File: assets/javascripts/discourse/templates/components/tc-message.hbs (hunk 3 above)
Comment (markvanlan) [reply assets/javascripts/discourse/templates/components/tc-message.hbs:32]: those are selectors to know which button was pressed, while passing in the `SHOW_LEFT` arg. I think it makes the JS much nicer.

File: assets/javascripts/discourse/components/tc-message.js (hunk 4 above)
Comment (pmusaraj) [reply assets/javascripts/discourse/components/tc-message.js:212]: Nevermind (<> not as fast with switches)

File: assets/javascripts/discourse/components/tc-message.js (hunk 5 above)
Comment (pmusaraj) [reply assets/javascripts/discourse/components/tc-message.js:261]: Ignore this too, pulled the repo locally, works quite well.

**Pull request merged**